src/Zigzag_persistence/include/gudhi/zigzag_persistence.h

```diff
@@ -0,0 +1,456 @@
+      : matrix_(
+            minNumberOfFaces,
+            [this](matrix_index columnIndex1, matrix_index columnIndex2) -> bool {
+              if (matrix_.get_column(columnIndex1).is_paired()) {
+                return matrix_.get_pivot(columnIndex1) < matrix_.get_pivot(columnIndex2);
+              }
+              return birthOrdering_.birth_order(births_.at(columnIndex1), births_.at(columnIndex2));
+            },
+            [this](matrix_index columnIndex1, matrix_index columnIndex2) -> bool { return false; }),
+        numArrow_(-1),
+        stream_interval_(std::move(stream_interval)) {}
+  /**
+   * @brief Updates the zigzag persistence diagram after the insertion of the given face.
+   *
+   * @tparam BoundaryRange Range type needing size, begin and end members.
+   * @param boundary Boundary of the inserted face. The boundary should be represented by all the faces with
+   * non-zero coefficients generating it. A face should be represented by the arrow number when the face appeared for
+   * the first time in the filtration (if a face was inserted and then removed and reinserted etc., only the last
+   * insertion counts). The face range should be ordered by increasing arrow numbers.
+   * @param dimension Dimension of the inserted face.
+   */
+  template <class BoundaryRange = std::initializer_list<index> >
+  void insert_face(const BoundaryRange& boundary, dimension_type dimension) {
+    ++numArrow_;
+    _process_forward_arrow(boundary, dimension);
+  }
+
+  /**
+   * @brief Updates the zigzag persistence diagram after the removal of the given face.
+   *
+   * @param arrowNumber Arrow number of when the face to remove was inserted for the last time.
+   * @param dimension Dimension of the face to remove.
+   */
+  void remove_face(index arrowNumber, dimension_type dimension) {
+    ++numArrow_;
+    _process_backward_arrow(arrowNumber, dimension);
+  }
+
+  /**
+   * @brief To use when a face is neither inserted nor removed, but the filtration moves along the identity operator
+   * on homology level. Useful to keep the birth/death indices aligned when insertions/removals are purposely skipped
+   * to avoid useless computation. Increases the arrow number by one.
+   */
+  void apply_identity() { ++numArrow_; }
+
+  /**
+   * @brief Outputs through the given callback method all birth indices which are currently not paired with 
+   * a death index.
+   * 
+   * @tparam F Type of the callback method. Takes two arguments: the dimension of the cycle and the birth index
+   * of the cycle.
+   * @param stream_infinite_interval Method processing the unpaired birth indices.
+   */
+  template <typename F>
+  void get_current_infinite_intervals(F&& stream_infinite_interval) {
+    for (auto& p : births_) {
+      if constexpr (erase_birth_history) {
+        auto& col = matrix_.get_column(p.first);
+        stream_infinite_interval(col.get_dimension(), p.second);
+      } else {
+        try {
```

Comment:
I could add to the matrix a method like `has_column(index)` which returns true if and only if the given index is not out of range. Otherwise, we can also hope that the benchmark with the oscillating rips branch shows that we can just keep `erase_birth_history` at true and then we would just remove this piece of code.

Reply:
`has_column` makes sense. It encourages making twice the same search in the hashtable though (has_column+get_column), so we could also consider a version of get_column that returns
* a `boost::optional<column&>`
* a `column*`, with nullptr to indicate out_of_range

Indeed, if `erase_birth_history` is just `true` in the future, that makes this point less relevant.